### PR TITLE
New module: Kernelized EDMD

### PR DESCRIPTION
**Pull request by sichinaga**
As promised, here is the implementation of Extended DMD (EDMD) [1] with the kernelized version of the method [2] backing the implementation for better-handling of high-dimensional snapshot data! Also included in this PR is the corresponding tutorial, which demonstrates how one might use the `EDMD` module in order to obtain an expanded set of eigenvalues and eigenfunctions from data. This module was a labor of love, and I look forward to any and all feedback from the team! :)

[1] M. O. Williams, I. G. Kevrekidis, and C. W. Rowley, A data-driven approximation of the koopman operator: extending dynamic mode decomposition, *J. Nonlinear Sci.*, **25** (2015), pp. 1307-1346. https://doi.org/10.1007/s00332-015-9258-5

[2] M. O. Williams, C. W. Rowley, and I. G. Kevrekidis, A kernel-based method for data-driven koopman spectral analysis, *J. Comput. Dynam.*, **2** (2015), pp. 247-265. https://doi.org/10.3934/jcd.2015005

### CODE REVIEW

File: pydmd/edmd.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+References:
+- M. O. Williams, C. W. Rowley, and I. G. Kevrekidis,
+A kernel-based method for data-driven koopman spectral analysis,
+J. Comput. Dynam., 2 (2015), pp. 247-265
+- M. O. Williams, I. G. Kevrekidis, and C. W. Rowley,
+A data-driven approximation of the koopman operator: extending
+dynamic mode decomposition, J. Nonlinear Sci., 25 (2015), pp. 1307-1346
+"""
+import numpy as np
+from sklearn.metrics.pairwise import pairwise_kernels
```
Comment (fandreuz) [inline pydmd/edmd.py:13]: Since you're importing it, I propose to add `scikit-learn` to the required packages in `setup.py`

File: pydmd/edmd.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    DMD operator for kernelized extended DMD.
+
+    :param svd_rank: the rank for the truncation; If positive integer, the
+        method uses the argument for the truncation; if float between 0 and 1,
+        the rank is the number of the biggest singular values that are needed
+        to reach the 'energy' specified by `svd_rank`; if -1, the method does
+        not compute truncation.
+    :type svd_rank: int or float
+    :param kernel_metric: the kernel function to apply.
+    :type kernel_metric: {"additive_chi2", "chi2", "linear",
```
Comment (fandreuz) [inline pydmd/edmd.py:30]: I think the available options should rather go in the `:param kernel_metric:` section above

File: pydmd/edmd.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+        # Obtain the eigenvalues and eigenvectors of K_hat.
+        # Note: The nonzero eigenvalues of K_hat are the DMD eigenvalues.
+        # DMD modes are derived via the left and right eigenvectors of K_hat.
+        eigenvalues, V_hat = np.linalg.eig(K_hat)
+        modes = np.linalg.multi_dot(
+            [X, Q, np.linalg.pinv(Sigma), np.linalg.inv(V_hat).T]
+        )
+
+        # Discard eigenvalue, mode pairs that correspond with zero eigenvalues.
```
Comment (fandreuz) [inline pydmd/edmd.py:110]:
```suggestion
        # Discard eigenvalues, mode pairs that correspond with zero eigenvalues.
```

File: pydmd/edmd.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        :param numpy.ndarray Gramian: the Gramian matrix.
+        :return: the (truncated) left-singular vectors matrix of the extended
+            feature matrix and the (truncated) singular values matrix of the
+            extended feature matrix.
+        :rtype: numpy.ndarray, numpy.ndarray
+        """
+        G_eigenvalues, G_eigenvectors = np.linalg.eig(Gramian)
+
+        # Note: The eigenvalues of the Gramian are the singular values of the
+        # feature embedding matrix, squared. Hence in theory, the eigenvalues
```
Comment (fandreuz) [inline pydmd/edmd.py:141]: The feature embedding matrix has some special properties? Correct me if I'm wrong, but squared singular values _in general_ can also have an imaginary part

File: pydmd/edmd.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        """
+        if 0 < self._svd_rank < 1:
+            cumulative_energy = np.cumsum(s**2 / (s**2).sum())
+            rank = np.searchsorted(cumulative_energy, self._svd_rank) + 1
+        elif self._svd_rank >= 1 and isinstance(self._svd_rank, int):
+            rank = min(self._svd_rank, len(s))
+        else:
+            rank = len(s)
+
+        return rank
```
Comment (fandreuz) [inline pydmd/edmd.py:178]:
```suggestion
            return np.searchsorted(cumulative_energy, self._svd_rank) + 1
        elif self._svd_rank >= 1 and isinstance(self._svd_rank, int):
            return min(self._svd_rank, len(s))
        else:
            return len(s)
```

File: pydmd/edmd.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        algorithm).
+        The reconstruction will generally be better in time instants near the
+        chosen snapshot; however increasing `opt` may lead to wrong results
+        when the system presents small eigenvalues. For this reason a manual
+        selection of the number of eigenvalues considered for the analyisis may
+        be needed (check `svd_rank`). Also setting `svd_rank` to a value
+        between 0 and 1 may give better results. Default is False.
+    :type opt: bool or int
+    :param kernel_metric: the kernel function to apply.
+    :type kernel_metric: {"additive_chi2", "chi2", "linear",
```
Comment (fandreuz) [inline pydmd/edmd.py:207]: Same as above

File: pydmd/edmd.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        :rtype: numpy.ndarray
+        """
+        if self._svd_modes is None:
+            raise ValueError("You need to call fit before")
+
+        if not isinstance(x, np.ndarray) or np.ndim(x) != 1:
+            raise ValueError("Input x must be a 1-D numpy array.")
+
+        K_xx = pairwise_kernels(
+            x[None, :],
```
Comment (fandreuz) [inline pydmd/edmd.py:262]:
```suggestion
            x[None],
```

File: pydmd/edmd.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        return np.linalg.multi_dot(
+            [
+                K_xx,
+                self._svd_modes,
+                np.linalg.pinv(np.diag(self.operator.svd_vals)),
+                self.operator.eigenvectors,
+            ]
+        )
+
+    def _test_kernel_inputs(self, kernel_metric, kernel_params):
```
Comment (fandreuz) [inline pydmd/edmd.py:277]:
```suggestion
    @staticmethod
    def _test_kernel_inputs(kernel_metric, kernel_params):
```

**Review by fandreuz — COMMENTED** [review]
Thanks for the new DMD @sichinaga! I left some minor comments

File: pydmd/edmd.py (hunk 3 above)
Comment (sichinaga) [reply pydmd/edmd.py:110]: This wording was actually intentional, as to refer to pairs of eigenvalues and modes as opposed to eigenvalues and modes separate from one another. Apologies that the wording is confusing 😅

File: pydmd/edmd.py (hunk 4 above)
Comment (sichinaga) [reply pydmd/edmd.py:141]: I believe singular values are always non-negative real numbers, in which case a squared singular value will also always be non-negative and real.

**Comment by sichinaga** [comment]
Thanks for reviewing the code @fandreuz !

The most recent commit should resolve everything. :)

File: pydmd/edmd.py (hunk 2 above)
Comment (sichinaga) [reply pydmd/edmd.py]: Just updated the documentation! Let me know what you think! :)

File: pydmd/edmd.py (hunk 1 above)
Comment (sichinaga) [reply pydmd/edmd.py:13]: Updated the `REQUIRED` list in `setup.py` to be
```
REQUIRED = ["numpy", "scipy", "matplotlib", "sklearn"]
```
Not sure if it matters if it's "sklearn" or "scikit-learn"?

File: pydmd/edmd.py (hunk 4 above)
Comment (fandreuz) [reply pydmd/edmd.py:142]: Ops, thanks :+1:

File: pydmd/edmd.py (hunk 1 above)
Comment (fandreuz) [reply pydmd/edmd.py:13]: Yeah it seems you can use both

File: pydmd/edmd.py (hunk 2 above)
Comment (fandreuz) [reply pydmd/edmd.py]: Thanks!

**Comment by sichinaga** [comment]
Alright! It looks like all of the tests are passing with the exception of a few very minor Codacy issues.

@fandreuz @mtezzele @ndem0 I think this PR is ready to merge unless anyone has any additional comments. :)

**Comment by mtezzele** [comment]
@sichinaga all ok, thanks! 
If you can just add the `rst` files to generate the doc of the new module then we can merge. Let me know if you prefer me to merge this one and open a PR just for the doc, instead.

**Comment by sichinaga** [comment]
> @sichinaga all ok, thanks! If you can just add the `rst` files to generate the doc of the new module then we can merge. Let me know if you prefer me to merge this one and open a PR just for the doc, instead.

@mtezzele if you could be the one to add the `rst` files, that would actually be really helpful. I'm not yet super familiar with adding the `rst` files, so this would probably be more efficient. Thank you so much!

**Comment by mtezzele** [comment]
> > @sichinaga all ok, thanks! If you can just add the `rst` files to generate the doc of the new module then we can merge. Let me know if you prefer me to merge this one and open a PR just for the doc, instead.
> 
> @mtezzele if you could be the one to add the `rst` files, that would actually be really helpful. I'm not yet super familiar with adding the `rst` files, so this would probably be more efficient. Thank you so much!

Sure no problem!

**Pull request merged**